ttnn/CMakeLists.txt

```diff
@@ -657,7 +697,92 @@ if(WITH_PYTHON_BINDINGS)
     )
 endif()
 
-add_library(ttnn SHARED ${TTNN_FINAL_SRC})
+function(use_precompiled_header TARGET)
+    if("${PRECOMPILED_HEADER_TARGET}" STREQUAL "")
+        set(PRECOMPILED_HEADER_TARGET "${TARGET}" PARENT_SCOPE)
+        target_precompile_headers(${TARGET} PUBLIC ${TTNN_PRECOMPILED_HEADERS})
+    else()
+        target_precompile_headers(${TARGET} REUSE_FROM ${PRECOMPILED_HEADER_TARGET})
+    endif()
+endfunction()
+
+set(PRECOMPILED_HEADER_TARGET "")
+
+function(add_ttnn_sublibrary SUBLIBRARY_NAME)
+    if(CMAKE_BUILD_TYPE STREQUAL "Debug")
```

Comment:
Can you share details on why is this if/else needed? 
And maybe capture in the comment in cmake itself. 
This part is surprising

Reply:
This speed up was achieved by using shared libraries for all the sub libraries in Debug build. 
For Release build, there is only one libraries so that there are not a lot of installation artifacts. There is no speed-up for release builds. 

I've updated the PR description too.

Reply:
Why not object libraries for both release and debug?

Why spend effort and add complexity optimizing debug build link time vs release?

Can we keep it simple?

It will be easier to maintain and refactor later if we keep it simple.

Reply:
If this is only for debug builds, I vote not to merge.

Reply:
Currently Debug takes almost 3 minutes for an incremental build, while Release takes around 28s. 

As an op writer, when we are in the change-build-test loop, waiting 3 mins for even a single line change is a significant impact on productivity.

Reply:
Is it really necessary to debug by changing one line at a time? Is gdb/lldb not helpful as an op writer? I’m honestly asking, because I haven’t had to do the low level op development yet.

I don’t think the complexity introduced here is worth the benefit. 

Can you keep the changes in a patch file, and apply them when you need them? Alternative a branch that you can continuously rebase?

Reply:
Detailed logging currently only works in Debug mode, which is necessary for most development or debugging work. 

gdb is useful to identify what potentially could be the problem. The only way to know for certain  is to fix it, re-compile and run. As an op writer, I end up making small changes and re-compiling a lot. 

If these changes are not merged into main, then the branch would quickly go out of date. Also if it's merged into main, all the other op writers would benefit from these changes immensely.

Reply:
Understood. Unfortunately, I am not supportive of the change which creates a larger difference between how debug and release builds are produced. While this helps to significantly improve the day-to-day for every op developer (very precious) this also creates situations when an issue only reproducible in release, but not in debug and now someone from the team will spend a very significant amount of time to figure out why and address such an issue. We must keep debug and release build as aligned as possible and find other opportunities to speedup the build.

Reply:
I would really like to find a solution that is acceptable, while also improving debug build time. I know this is a significant pain point for the Conv team. 

Three approaches I could think of are : 
1. Using shared libraries for the sub libraries is controlled by an option instead of build type, This is off by default to ensure debug and release builds are built the same way. 
2. Use shared libraries for all build types.
3. Merge this PR with only object libraries, and let the developer patch CMakeLists to use shared libraries.

Reply:
If this PR must be merged for productivity, I would support option 1 or 3. 1 sounds like it would be best.

I don’t like option 2, because it’s hard for distribution, and there is the risk of linking stale/incompatible libraries.

I hope we can find other solutions to the slow link stage that don’t require shared libraries. I would like to take a look at this when I am back from vacation in 2025.

Do we understand why shared libraries are helping vs object libraries?

Reply:
> Do we understand why shared libraries are helping vs object libraries?

Shared libs are the only ones that are linked (other than executables).  Static libs and Object libs are not linked.  Static are archived all together (think: tar), but no linking is done.  Object forego even the archiving and just bubble up the individual pieces to the next layer.

Making shared libs along the way is a divide-and-conquor approach.  Each link is a subset of the whole which (a) cuts down on the amount of work that needs to be done and (b) can be farmed out to multiple cores.  A single link is going to spin a single core no matter how many CPUs you have.  Also with all the un-weeded out bits of all the .o files, there's going to be a massive amount of duplication that it needs to sift through.  I looked yesterday and any file that `#include <host_api.h>` is getting 10818 lines of code added to its TU.  Some will be stripped by the pre-processor, and some symbols should be dropped in the making of the .o.  But there will be many symbols that remain.  Those all need to be de-duplicated by the linker.  Meanwhile they're consuming RAM and if you run out and touch swap, then the performance will tank further.

TTNN also includes other headers from Metalium byond host_api.h.  The total set comes to 17554 LoC.  And that's not including anything from TTNN or any other dependency.

Ways to address the issue:
* Fix APIs -- don't dump things in public headers that don't need to be there.  This avoids a fanout effect of pointless work by downstream consumers.  This is also a very long term project.  But we should be mindful of in our day-to-day.
* Split up into piles of .so.  This speeds up linking because it divides and conquers and can use multiple CPUs.  This also risks ODR violations when combining all those .so's back together into a consuming product.  It also prevents LTO (whether we want that is a tradeoff between performance vs link time).
* Use a more efficient linker.  One that can handle bigger inputs, and is more efficient with deduplicating symbols.  There's chatter the other day about speedups from using `mold` over `ld`.  CMake has native support for a number of linkers.  @sankarmanoj-tt how much performance boost do you see by using CMake v3.29+ (upgrade if needed) and then setting https://cmake.org/cmake/help/latest/variable/CMAKE_LINKER_TYPE.html to GOLD or MOLD (will also need to install each linker, naturally).

Reply:
> don't dump things in public headers that don't need to be there

I want a T-shirt that says this.

Thanks for the overview here.

I submitted a couple PRs that pulled things out of public header files. Sounds like a lot more cleanup there.

Reply:
I've implemented option 1, i.e. adding an CMake option called `ENABLE_TTNN_SHARED_SUBLIBS`